staging/src/k8s.io/dynamic-resource-allocation/controller/controller.go

```diff
@@ -517,50 +525,87 @@ func (ctrl *controller) getParameters(ctx context.Context, claim *resourcev1alph
+	for _, claim := range claims {
+		if claim.Claim.Status.Allocation != nil {
+			// This can happen when two PodSchedulingContext objects trigger
+			// allocation attempts (first one wins) or when we see the
+			// update of the PodSchedulingContext object.
+			logger.V(5).Info("Claim is already allocated, skipping allocation", "claim", claim.PodClaimName)
+			continue
+		}
+		needAllocation = append(needAllocation, claim)
+	}
+
+	if len(needAllocation) == 0 {
+		logger.V(5).Info("No claims need allocation, nothing to do")
 		return nil
 	}
 
-	claim = claim.DeepCopy()
-	if !ctrl.hasFinalizer(claim) {
-		// Set finalizer before doing anything. We continue with the updated claim.
-		logger.V(5).Info("Adding finalizer")
-		claim.Finalizers = append(claim.Finalizers, ctrl.finalizer)
-		var err error
-		claim, err = ctrl.kubeClient.ResourceV1alpha2().ResourceClaims(claim.Namespace).Update(ctx, claim, metav1.UpdateOptions{})
-		if err != nil {
-			return fmt.Errorf("add finalizer: %v", err)
+	for _, claimAllocation := range needAllocation {
+		if !ctrl.hasFinalizer(claimAllocation.Claim) {
+			patchFn := func() error {
+				claim := claimAllocation.Claim.DeepCopy()
+				// Set finalizer before doing anything. We continue with the updated claim.
+				logger.V(5).Info("Adding finalizer", "claim", claim.Name)
+				claim.Finalizers = append(claim.Finalizers, ctrl.finalizer)
+				var err error
+				claim, err = ctrl.kubeClient.ResourceV1alpha2().ResourceClaims(claim.Namespace).Update(ctx, claim, metav1.UpdateOptions{})
+				if err != nil {
+					return fmt.Errorf("add finalizer: %v", err)
+				}
+				ctrl.claimCache.Mutation(claim)
+				claimAllocation.Claim = claim
+				return nil
+			}
+			err := retry.RetryOnConflict(retry.DefaultRetry, patchFn)
+			if err != nil {
+				return err
+			}
 		}
-		ctrl.claimCache.Mutation(claim)
 	}
 
 	logger.V(5).Info("Allocating")
-	allocation, err := ctrl.driver.Allocate(ctx, claim, claimParameters, class, classParameters, selectedNode)
+	allocations, err := ctrl.driver.Allocate(ctx, claims, selectedNode)
 	if err != nil {
 		return fmt.Errorf("allocate: %v", err)
 	}
-	claim.Status.Allocation = allocation
-	claim.Status.DriverName = ctrl.name
-	if selectedUser != nil {
-		claim.Status.ReservedFor = append(claim.Status.ReservedFor, *selectedUser)
+
+	if len(claims) != len(allocations) {
+		logger.Error(nil, "resource driver returned less allocation results than resource claims were requested")
+		return fmt.Errorf("allocationResults are fewer than resourceClaims")
 	}
-	logger.V(6).Info("Updating claim after allocation", "claim", claim)
-	claim, err = ctrl.kubeClient.ResourceV1alpha2().ResourceClaims(claim.Namespace).UpdateStatus(ctx, claim, metav1.UpdateOptions{})
-	if err != nil {
-		return fmt.Errorf("add allocation: %v", err)
+
+	// Update claims' status with allocation info
+	for caIdx, allocation := range allocations {
+		patchFn := func() error {
+			claim := claims[caIdx].Claim
+			if err != nil {
```

Comment:
Which `err` is being checked here? This looks odd.

Reply:
Oof, yes, this is quite an artifact, apologies.

Reply:
This is actually one more argument in favor of `if err := ...; err != nil {}`: the `err` instance is not visible to code coming later and thus cannot be used accidentally.

It wouldn't have helped here because `err` comes from a function with two return values, but in principle the argument about reduced visibility applies.

Reply:
That's true.